crates/chia-sdk-driver/src/layers/p2_eip712_message_layer.rs

```diff
@@ -0,0 +1,298 @@
+                pubkey: self.pubkey.to_vec().into(),
+            },
+        };
+        ctx.alloc(&curried)
+    }
+
+    fn construct_solution(
+        &self,
+        ctx: &mut SpendContext,
+        solution: Self::Solution,
+    ) -> Result<NodePtr, DriverError> {
+        ctx.alloc(&solution)
+    }
+
+    fn parse_puzzle(_allocator: &Allocator, _puzzle: Puzzle) -> Result<Option<Self>, DriverError> {
+        Ok(None)
+    }
+
+    fn parse_solution(
+        allocator: &Allocator,
+        solution: NodePtr,
+    ) -> Result<Self::Solution, DriverError> {
+        Ok(P2Eip712MessageSolution::from_clvm(allocator, solution)?)
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use crate::assert_puzzle_hash;
+
+    use super::*;
+    use chia_consensus::consensus_constants::TEST_CONSTANTS;
+    use chia_protocol::Bytes;
+    use chia_sdk_test::Simulator;
+    use chia_sdk_types::Conditions;
+    use clvm_traits::clvm_quote;
+    use clvm_utils::ToTreeHash;
+    use clvmr::chia_dialect::ENABLE_KECCAK_OPS_OUTSIDE_GUARD;
+    use clvmr::reduction::Reduction;
+    use clvmr::serde::node_from_bytes;
+    use ethers::core::rand::thread_rng;
+    use ethers::prelude::*;
+    use ethers::signers::LocalWallet;
+    use k256::ecdsa::signature::hazmat::PrehashVerifier;
+    use k256::ecdsa::{Signature as K1Signature, SigningKey, VerifyingKey as K1VerifyingKey};
+    use rstest::rstest;
+
+    #[test]
+    fn test_puzzle_hash() -> anyhow::Result<()> {
+        assert_puzzle_hash!(P2_EIP712_MESSAGE_PUZZLE => P2_EIP712_MESSAGE_PUZZLE_HASH);
+        Ok(())
+    }
+
+    #[test]
+    fn test_softfork_cost() -> anyhow::Result<()> {
+        let ctx = &mut SpendContext::new();
+        let puzzle_bytes =
+            hex!("ff02ffff03ffff09ffff3eff02ffff3eff05ff0bff178080ff2f80ff80ffff01ff088080ff0180");
+
+        let puzzle_ptr = node_from_bytes(&mut ctx.allocator, puzzle_bytes.as_slice())?;
+        let solution_ptr = vec![
+            // warning: old domain separator w/o version; do NOT use!
+            Bytes::new(
+                hex!("1901098ccd7d09a29365582c3f7590712bc2c2eb8503586f8a4c628c61c73ffbe4aa")
+                    .to_vec(),
+            ), // PREFIX_AND_DOMAIN_SEPARATOR
+            Bytes::new(
+                hex!("72930978f119c79f9de7a13bd50c9b3261132d7b4819bdf0d3ca4d4c37ade070").to_vec(),
+            ), // TYPE_HASH
+            Bytes::new(
+                hex!("5c777c45fd52a17a54e420742cadc56172847d9a106ff0ff8af38ef757d84829").to_vec(),
+            ), // my_id
+            Bytes::new(
+                hex!("d842dfa1453a130a8be66bc32708a2d1884662d7daaa4aae530be3259fa6712f").to_vec(),
+            ), // delegated_puzzle_hash
+            Bytes::new(
+                hex!("9f61fdf6077c3eeb96eaa4dd450b11ba3ae17746a2c304388218137972c7ba4c").to_vec(),
+            ), // signed_hash
+        ]
+        .to_clvm(&mut ctx.allocator)?;
+
+        let Reduction(cost, _) = clvmr::run_program(
```

Comment:
There's a `run_puzzle` utility, maybe it would be good to change it to return the cost also or add another `run_puzzle_with_cost` helper. The main reasoning is to consolidate the places where flags are defined. Maybe you need custom flags here though.

Reply:
I'll do for a while (`ENABLE_KECCAK_OPS_OUTSIDE_GUARD` until next hard fork)